lrpar/src/lib/ctbuilder.rs

```diff
@@ -1042,161 +1042,164 @@ where
     }
 
     /// Generate the wrappers that call user actions
-    fn gen_wrappers(&self, grm: &YaccGrammar<StorageT>) -> String {
-        let mut outs = String::new();
-
-        outs.push_str("\n\n    // Wrappers\n\n");
-
-        let (parse_paramname, parse_paramdef) = match grm.parse_param() {
-            Some((name, tyname)) => (name.to_owned(), format!("{}: {}", name, tyname)),
-            None => ("()".to_owned(), "_: ()".to_owned()),
+    fn gen_wrappers(
+        &self,
+        grm: &YaccGrammar<StorageT>,
+    ) -> Result<TokenStream, proc_macro2::LexError> {
+        let (parse_param_unit, parse_paramname, parse_paramdef);
+        match grm.parse_param() {
+            Some((name, tyname)) => {
+                parse_param_unit = tyname.trim() == "()";
+                parse_paramname = str::parse::<TokenStream>(name)?;
+                let ty = str::parse::<TokenStream>(tyname)?;
+                parse_paramdef = quote!(#parse_paramname: #ty);
+            }
+            None => {
+                parse_param_unit = true;
+                parse_paramname = quote!(());
+                parse_paramdef = quote! {_: ()};
+            }
         };
+
+        let mut wrappers = TokenStream::new();
         for pidx in grm.iter_pidxs() {
             let ridx = grm.prod_to_rule(pidx);
-            let pidx_num = UnsuffixedUsize(usize::from(pidx));
 
             // Iterate over all $-arguments and replace them with their respective
             // element from the argument vector (e.g. $1 is replaced by args[0]). At
             // the same time extract &str from tokens and actiontype from nonterminals.
-            write!(outs,
-                "    fn {prefix}wrapper_{}<'lexer, 'input: 'lexer>({prefix}ridx: ::cfgrammar::RIdx<{storaget}>,
-                      {prefix}lexer: &'lexer dyn ::lrpar::NonStreamingLexer<'input, {lexertypest}>,
-                      {prefix}span: ::cfgrammar::Span,
-                      mut {prefix}args: ::std::vec::Drain<{edition_lifetime} ::lrpar::parser::AStackType<<{lexertypest} as ::lrpar::LexerTypes>::LexemeT, {actionskind}<'input>>>,
-                      {parse_paramdef})
-                   -> {actionskind}<'input> {{",
-                quote!(#pidx_num),
-                storaget = type_name::<StorageT>(),
-                lexertypest = type_name::<LexerTypesT>(),
-                prefix = ACTION_PREFIX,
-                parse_paramdef = parse_paramdef,
-                actionskind = ACTIONS_KIND,
-                edition_lifetime = if self.rust_edition != RustEdition::Rust2015 { "'_, " } else { "" },
-            ).ok();
-
+            let wrapper_fn = format_ident!("{}wrapper_{}", ACTION_PREFIX, usize::from(pidx));
```

Comment:
When writing all of this, I had kind of been under the notion that
`format_ident!()` didn't support `{variable}` style of format strings.

But that is not actually the case, that type of format string appears to work fine,
I'm not sure what it was that caused me to believe that.

I tend to feel that the main places that needed `{variables}` were where we had lots of `{}` variables
which are now mostly replaced with `quotes`, which have `#variables` instead.